### ISSUE TITLE
When Geofence is present removing a gridset makes the layer non accesible anymore

### ISSUE THREAD

**Issue by nmco** [geosolutions-it/geoserver#225]
This was reprove din this ticket:
https://geosolutions.freshdesk.com/a/tickets/46919

The client fixed this by removing GeoFence, deleting the layer and then reinstall GeoFence ...

@taba90 **let's spend 30 MAX** trying to reproduce and fi that's the case open a JIRA ticket.

**Comment by nmco** [comment]
Let's follow the usual workflow:

1. Time-box the initial investigation of the issue to 4 hours MAXIMUM.
2. Clarify the context of the task, make sure you have everything needed to move on.
3. Provide a plan and estimate to handle the reported issue, report on the issue.
4. Wait for the green light from the PM before moving on with the implementation.

**Comment by zvrablik** [comment]
@nmco I run out of time. Should I continue?
What I have done:
1. geofence setup, followed one tutorial
2. Marco helped me to reproduce the issue as I don't have access to freshdesk to see the bug description content
3. I see there is dynamic proxy around SecuredFeatureTypeInfo and it fails.

**Comment by nmco** [comment]
Go ahead spend more two hours if needed.

**Comment by zvrablik** [comment]
Steps to reproduce:
1. start geoserver with configured GeoFence ( remember to add GWC property to geofence properties file)
2. go to Tiled layers pick one layer remove gridset eg. the mercator. Select apply button
3. go to layers

Exception happens [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java#L703-L703)

The problem is that the SecuredFeatureTypeInfo is wrapped with ModificationProxy. This proxy is not type of FeatureTypeInfo to get processed [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java#L662-L662) but only ResourceInfo.

Because the type is created [here](https://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxy.java#L144-L152) and takes the ResourceInfo as class to proxy.

All in the if statements [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java#L661-L662) has inherited the ResourceInfo so using ResourceInfo instead of FeatureTypeInfo is not possible.

The type is not used for anything but checking the type I would like to check if the info is proxy and unwrap it to be able to check the info object type without ModificationProxy.

It was one liner so I tested it and than I got another issue a bit later [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/main/src/main/java/org/geoserver/security/decorators/SecuredLayerInfo.java#L31-L38) 
The issue is the SecuredLayerInfo has already SecuredFeatureTypeInfo as resource wrapping the FeatureTypeInfoImp.

@aaime should I take this as part of this bug report? I have not investigated much, but it is related to the bug reported. Same steps to reproduce, but I run out of assigned time to continue investigation.

in case I should leave the other issue to another bug report my plan is 

1. unwrap the proxy before the instanceof checks [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/extension/geofence/src/main/java/org/geoserver/geofence/GeofenceAccessManager.java#L661-L663) , add integration test to GeofenceAccessManager with Layer having resource wrapped by ModificationProxy.
2. manual testing
3. backports to 2.18 and 2.19 branches

estimates 3 hours

@aaime  Do you agree with my plan? What about the subsequent issue?

**Comment by aaime** [comment]
@zvrablik unwrapping proxies should be the last resort, normally code that does that is aware of them because it's involved also in the creation and management of the same. The rest of the code should be oblivious to them, as unwrapping breaks the functionality the proxy was set up for (in particular modification proxy prevents config objects to be modified without calling "save" on the catalog).

The security subsytem should not know about ModificationProxy... is it possible to modify the creation of the modification proxy so that it is made aware of the interfaces implemented by the returned object in particular, see if it's implemneting a sub-interface of "getReturnedType" and use that one instead?

**Comment by zvrablik** [comment]
The proxy is created [here](http://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/main/src/main/java/org/geoserver/catalog/impl/ModificationProxy.java#L144-L150) 
The method is getResource and its return type is ResourceInfo.
Hmm, do you think I could use instead of method.getReturnType() the result.getClass() in ModificationProxy?

**Comment by aaime** [comment]
@zvrablik I've already suggested an algorithm above: "made aware of the interfaces implemented by the returned object in particular, see if it's implemneting a sub-interface of "getReturnedType" and use that one instead".

Using the class will not work in general, I believe Java proxies work only with interfaces. I was suggesting to try and use a more focused interface... but I have to admit I am stumped, the creation of the proxy eventually [leads here](https://github.com/geoserver/geoserver/blob/5afb16b47e42dd4594a3ceba834076ff79733c71/src/main/src/main/java/org/geoserver/catalog/impl/ProxyUtils.java#L86), which should already be making the proxy implement all interfaces already implemented by the object being wrapped.... I'm guessing something is amiss in there, I would expect it to also add, among the others, FeatureTypeInfo.... can you check? Maybe that's the code which needs fixing.

**Comment by zvrablik** [comment]
yes, proxies can work only with interfaces.

Javadoc of the getInterfaces
```
Returns the interfaces directly implemented by the class or interface represented by this object.
If this object represents a class, the return value is an array containing objects representing all
```
code [here](https://github.com/geoserver/geoserver/blob/97f2445718ae01fdecf2dc8f0d1bd125734b4161/src/main/src/main/java/org/geoserver/catalog/impl/ProxyUtils.java#L86-L87) returns empty list when requested list of interfaces ... that is reason why the ResourceInfo is used instead.

It has to iterate over all parent classes and collect all interfaces implemented in parent classes - could this cause performance issues ? I see a lot of proxies. Maybe there could be added caching hashmap to do all this lookup only once. Picking one of multiple interfaces may be tricky. Other way - but less reliable is ensure all leaf classes implement the requested interface.
Eg. SecuredFeatureTypeInfo implements directly FeatureTypeInfo, but it is easy to forget as parent already has the interface ...

WDYT?

**Comment by aaime** [comment]
I like the first alternative better, caching the list of interfaces. There is already a cache in that class, "PROXY_CLASS_CACHE"... given that the set of classes being proxied is small, a ConcurrentHashMap should work fine. In cases where you need a cache for a potentially large set of objects, considering using a SoftValueHashMap or a WeakValueHashMap from GeoTools instead.

**Comment by zvrablik** [comment]
ok, 
I am going to iterate over all parents and I will use the cache. I agree with ConcurrentHashMap, should be enough.
changing estimates to **4 h** because of integration test

**Comment by aaime** [comment]
This fix came out for another reason, but it affects the same line: https://github.com/geoserver/geoserver/pull/5077/files
It's likely going to solve the same problem.
